plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php

```diff
@@ -1688,4 +1799,80 @@ protected function adjust_price_filter_for_tax_class( $price_filter, $tax_class
 
 		return $price_filter + array_sum( $taxes );
 	}
+
+	/**
+	 * Registers handlers for a collection.
+	 *
+	 * @param string        $collection_name The name of the custom collection.
+	 * @param callable      $build_query     A hook returning any custom query arguments to merge with the collection's query.
+	 * @param callable|null $frontend_args   An optional hook that returns any frontend collection arguments to pass to the query builder.
+	 * @param callable|null $editor_args     An optional hook that returns any REST collection arguments to pass to the query builder.
+	 * @param callable|null $preview_query   An optional hook that returns a query to use in preview mode.
+	 *
+	 * @throws \InvalidArgumentException If collection handlers are already registered for the given collection name.
+	 */
+	protected function register_collection_handlers( $collection_name, $build_query, $frontend_args = null, $editor_args = null, $preview_query = null ) {
```

Comment:
@ObliviousHarmony Given the environment we are currently working in (a single huge PHP class), this abstraction is great! 

However, I feel that we are assuming that the editor preview will always have the exact same needs as the frontend. While this is true in most cases, I can't help but see that this is a bit more complex than necessary. This is because the flow introduces one more logical leap: some callbacks will be fed into arguments into others, effectively introducing another layer.

What if we only have three arguments here? (a) build_frontend_query, (b) build_editor_query, (c - optional) build the preview query?

With that, I can already see @kmanijak's's [pull request on upsells](https://github.com/woocommerce/woocommerce/pull/51374/files#diff-933825fdae3edb1301562e70acef6627856dd1c39c7475056ef21ac4f0ee9941R1917) repeating the same logic to grab the "collection arguments" while we could introduce a few helpers in the ProductCollectionUtils class.

For example, instead of:
```php
$this->register_collection_handlers(
    build_query_fn,
    get_frontend_args,
    get_editor_args,
    get_preview_args,
);
```
Which will end up repeating much of the code for every collection that uses a "viewed" entity to:
```php
$this->register_collection_handlers(
    build_frontend_query,
    build_editor_query,
    build_preview_query (optional)
);
```

And at the same time within the `ProductCollectionUtils` class:
```php
public function get_viewed_product( $block ) {
	$location = $block->context['productCollectionLocation'] ?? false;
	if ( ! $location || ! is_array( $location ) || ! isset( $location['type'] ) || 'product' !== $location['type'] || ! isset( $location['sourceData'] ) || empty( $location['sourceData']['productId'] ) ) {
		return false;
	}

	$product = wc_get_product( $location['sourceData']['productId'] );
	if ( ! ( $product instanceof WC_Product ) ) {
		return false;
	}

	return $product;
}
```

and a similar for the editor side, like:

```php
public function rest_get_viewed_product( WP_REST_Request ) {
    //...
    return false || WC_Product;
}
```

That way this case would be:
```php
$this->register_collection_handlers(
    function( $query, $block ) {
        $product = ProductCollectionUtils::get_viewed_product( $block );
        if ( ! $product ) {
            return $query; // or prevent results.
        }
        
        return array(
            'post__in' => wc_get_related_products(
            $product->get_id(),
            100
       ),
    function( $query, $request ) {
        $product = ProductCollectionUtils::rest_get_viewed_product( $request );
        if ( ! $product ) {
            return $query; // or prevent results.
        }
        
        return array(
            'post__in' => wc_get_related_products(
            $product->get_id(),
            100
       ),
    );
}
```

What do you think? These helpers can later be used by 3PDs as well, I guess, or be a concrete example of how to use "viewed" entities in a standard way. 

I don't have a strong preference, just sharing my vision here. Feel free to question my thinking :)

Reply:
I was also reviewing "Upsells" collection and I also noticed that we are repeating same logic. I don't have strong opinion for signature of `register_collection_handlers` but I think it would great to have utility functions in `ProductCollectionUtils` class. Although, I don't think this needs to be in scope of this PR. We can probably create a new issue for this and merge this PR 🙂

Reply:
While this is a great discussion @xristos3490, I think we should have this in the discussion thread related to extensibility. At least in our case now, it's okay because this is all internal and subject to change.

Reply:
As for moving it to `ProductCollectionUtils`, I think we should have a more specific issue around refactoring the `ProductCollection` class. Dumping everything in a utils class isn't really any better and I think there's a lot of interesting ways we could refactor the collection class. At least for now, I'd like to leave this here.

Reply:
> While this is a great discussion @xristos3490, I think we should have this in the discussion thread related to extensibility. At least in our case now, it's okay because this is all internal and subject to change.

Agreed! I really like the suggestion from @xristos3490. But It makes sense to not block this PR from merging. We can continue the discussion in the [GH discussion](https://github.com/woocommerce/woocommerce/discussions/51018) and refactor the code if needed once we have some agreement on the extensibility points 🙂

Reply:
Thank you, everyone! We don't need to solve everything here; I agree! This class definitely needs a complete refactor. I've added [these thoughts](https://github.com/woocommerce/woocommerce/discussions/51018#discussioncomment-10680769) to the discussion.

Reply:
> As for moving it to ProductCollectionUtils, I think we should have a more specific issue around refactoring the ProductCollection class. Dumping everything in a utils class isn't really any better and I think there's a lot of interesting ways we could refactor the collection class. At least for now, I'd like to leave this here.

Makes sense to me. I have created a new issue to refactor `ProductCollection` class -- #51534